lib/spack/spack/environment.py

```diff
@@ -374,16 +402,41 @@ def __init__(self, path, init_file=None):
     def _read_manifest(self, f):
         """Read manifest file and set up user specs."""
         self.yaml = _read_yaml(f)
+
+        self.read_specs = OrderedDict()
+
+        for item in list(self.yaml.values())[0].get('definitions', []):
+            entry = copy.deepcopy(item)
+            when = _eval_conditional(entry.pop('when', 'True'))
+            assert len(entry) == 1
+            if when:
+                name, spec_list = list(entry.items())[0]
+                user_specs = SpecList(name, [s for s in spec_list if s],
```

Comment:
If someone puts an empty string into their yaml. If you think that's an unnecessary level of input sanitization I can simplify it.

Reply:
I kind of do but mainly because I don't like the input sanitization in the middle of the rest of the logic. I think more generic sanitization function applied to the configuration as it is read would be nice though (and wouldn't interfere with understanding of the logic here).

Reply:
Changed.